### PR TITLE
feat(zink): add order-based #[storage] macro

### PR DESCRIPTION
**Pull request by 41rjordan**
added #[storage] proc macro in /zink/codegen, storage keys are derived based on variable order (e.g first #[storage] in project will take 0x0 storage key, second will take 0x1 storage key, etc)

only i32s are supported, maybe it's worth making ToStorage trait or something like that later, but before that we should probably change sstore/sload i/os to 32 byte words for convenience (now they're all i32 and macro works with i32 keys)

### example

```rust
#[storage]
pub type Counter = i32;
```

turns into

```rust
struct Counter;

impl zink::Storage<i32> for Counter {
    const STORAGE_KEY: i32 = 0i32;

    fn get() -> i32 {
        zink::ffi::evm::sload(Self::STORAGE_KEY)
    }

    fn set(value: i32) {
        zink::ffi::evm::sstore(Self::STORAGE_KEY, value);
    }
}
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -57,4 +57,4 @@ zint = { path = "zint", version = "=0.1.3" }
 
 [profile]
 dev = { panic = "abort"}
-release = { panic = "unwind" }
+release = { panic = "unwind" }
```
Comment (clearloop) [inline Cargo.toml:60]: plz create a new line after this line or there will be a EOF warning in github

same in other files

File: .idea/.gitignore [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (clearloop) [inline .idea/.gitignore:1]: please ignore `.idea` in `zink/.gitignore`

File: cli/src/new.rs [hunk 3]
```diff
@@ -81,7 +81,7 @@ impl New {
         let src = self.path.join("src");
         fs::create_dir_all(&src)?;
         fs::write(
-            src.join("lib.rs"),
+            src.join("mod"),
```
Comment (clearloop) [inline cli/src/new.rs:84]: hmmm why changing it to `mod`?

this command creates a new template zink project from rust library, and the root of the library should be `lib.rs`

File: zink/codegen/src/lib.rs [hunk 4]
```diff
@@ -34,3 +37,9 @@ pub fn event(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input as DeriveInput);
     event::parse(input)
 }
+
```
Comment (clearloop) [inline zink/codegen/src/lib.rs:40]: please provide docs for this proc-macro, for example https://github.com/zink-lang/zink/blob/main/zink/codegen/src/lib.rs#L8, and the generated docs https://docs.zink-lang.org/rustdocs/zink_codegen/derive.Event.html

File: zink/src/lib.rs [hunk 5]
```diff
@@ -2,11 +2,13 @@
...
 #![no_std]
 
+extern crate alloc;
+
 mod event;
 pub mod ffi;
 
 pub use self::event::Event;
-pub use zink_codegen::Event;
+pub use zink_codegen::Storage;
```
Comment (clearloop) [inline zink/src/lib.rs:11]:
```suggestion
pub use zink_codegen::{Event, storage};
```

File: zink/codegen/Cargo.toml [hunk 6]
```diff
@@ -9,3 +9,5 @@ proc-macro = true
 [dependencies]
 syn.workspace = true
 quote.workspace = true
+sha3 = "0.10.8"
```
Comment (clearloop) [inline zink/codegen/Cargo.toml:12]: we can remove this dependency since it is currently not using?

File: zink/codegen/src/storage.rs [hunk 7]
```diff
@@ -0,0 +1,71 @@
+extern crate proc_macro;
+
+use proc_macro2::TokenStream;
+use std::any::Any;
+use std::sync::atomic::AtomicI32;
+use std::sync::atomic::Ordering::Relaxed;
+use quote::{quote, ToTokens};
+use syn::ItemType;
+
+pub trait Storage<T> {
```
Comment (clearloop) [inline zink/codegen/src/storage.rs:10]: please provide simple docs for this `trait`

**Review by clearloop — COMMENTED** [review]
really appreciate ur work! btw we can update the storage example https://github.com/zink-lang/zink/blob/main/examples/storage/src/lib.rs#L13 with the new proc-macro now

File: cli/src/new.rs (hunk 3 above)
Comment (41rjordan) [reply cli/src/new.rs:84]: lol new ide sry

**Comment by clearloop** [comment]
Things I've changed: 

1. the title of this PR: https://www.conventionalcommits.org/en/v1.0.0/
2. some outdated code, removed them with `git rm -rf .idea`
3. some nits in the new introduced `Storage` trait that makes compilation failed

**Pull request merged**